core/vm/actor/builtin/miner/miner_actor.cpp

```diff
@@ -5,9 +5,883 @@
+    if (runtime.getCurrentEpoch() > state.post_state.proving_period_start
+                                        + kWindowedPostChallengeDuration) {
+      return VMExitCode::MINER_ACTOR_POST_TOO_LATE;
+    }
+    if (runtime.getCurrentEpoch() <= state.post_state.proving_period_start) {
+      return VMExitCode::MINER_ACTOR_POST_TOO_EARLY;
+    }
+    OUTCOME_TRY(verifyWindowedPoSt(runtime, state, params2));
+    state.post_state.num_consecutive_failures = 0;
+    state.post_state.proving_period_start += kProvingPeriod;
+    state.proving_set = state.sectors;
+    OUTCOME_TRY(runtime.commitState(state));
+    OUTCOME_TRY(runtime.send(
+        kStoragePowerAddress, kOnMinerSurprisePoStSuccessMethodNumber, {}, 0));
+    return outcome::success();
+  }
+
+  ACTOR_METHOD(onDeleteMiner) {
+    if (runtime.getImmediateCaller() != kStoragePowerAddress) {
+      return VMExitCode::MINER_ACTOR_WRONG_CALLER;
+    }
+    OUTCOME_TRY(runtime.deleteActor());
+    return outcome::success();
+  }
+
+  ACTOR_METHOD(preCommitSector) {
+    OUTCOME_TRY(params2, decodeActorParams<PreCommitSectorParams>(params));
+    OUTCOME_TRY(state, assertCallerIsWorker(runtime));
+
+    OUTCOME_TRY(already_commited,
+                Amt(runtime.getIpfsDatastore(), state.sectors)
+                    .contains(params2.sector));
+    if (already_commited) {
+      return VMExitCode::MINER_ACTOR_ILLEGAL_ARGUMENT;
+    }
+
+    auto deposit = precommitDeposit(
+        state.info.sector_size, params2.expiration - runtime.getCurrentEpoch());
+    OUTCOME_TRY(confirmPaymentAndRefundChange(runtime, deposit));
+
+    Hamt hamt_precommitted_sectors{runtime.getIpfsDatastore(),
+                                   state.precommitted_sectors};
+    OUTCOME_TRY(hamt_precommitted_sectors.setCbor(
+        uvarintKey(params2.sector),
+        SectorPreCommitOnChainInfo{
+            .info = params2,
+            .precommit_deposit = deposit,
+            .precommit_epoch = runtime.getCurrentEpoch(),
+        }));
+    OUTCOME_TRY(hamt_precommitted_sectors.flush());
+    state.precommitted_sectors = hamt_precommitted_sectors.cid();
+    OUTCOME_TRY(runtime.commitState(state));
+
+    if (params2.expiration <= runtime.getCurrentEpoch()) {
+      return VMExitCode::MINER_ACTOR_ILLEGAL_ARGUMENT;
+    }
+
+    OUTCOME_TRY(duration, maxSealDuration(params2.registered_proof));
+    OUTCOME_TRY(
+        enrollCronEvent(runtime,
+                        runtime.getCurrentEpoch() + duration + 1,
+                        {
+                            .event_type = CronEventType::PreCommitExpiry,
+                            .sectors = RleBitset{params2.sector},
+                            .registered_proof = params2.registered_proof,
+                        }));
+
+    return outcome::success();
+  }
+
+  ACTOR_METHOD(proveCommitSector) {
+    OUTCOME_TRY(params2, decodeActorParams<ProveCommitSectorParams>(params));
+    OUTCOME_TRY(state, assertCallerIsWorker(runtime));
```

Comment:
It has been changed https://github.com/filecoin-project/specs-actors/blob/ad3d104d72523b18d6eccd9db140f43a00eaea3c/actors/builtin/miner/miner_actor.go#L277